omnigibson/systems/macro_particle_system.py

```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
+            positions,
+            velocities=None,
+            angular_velocities=None,
+            orientations=None,
+            scales=None,
+            **kwargs,
+    ):
+        """
+        Generates new particles
+
+        Args:
+            positions (np.array): (n_particles, 3) shaped array specifying per-particle (x,y,z) positions
+            velocities (None or np.array): (n_particles, 3) shaped array specifying per-particle (x,y,z) velocities.
+                If not specified, all will be set to 0
+            angular_velocities (None or np.array): (n_particles, 3) shaped array specifying per-particle (ax,ay,az)
+                angular velocities. If not specified, all will be set to 0
+            orientations (None or np.array): (n_particles, 4) shaped array specifying per-particle (x,y,z,w) quaternion
+                orientations. If not specified, all will be sampled randomly
+            scales (None or np.array): (n_particles, 3) shaped array specifying per-particle (x,y,z) scales.
+                If not specified, will be uniformly randomly sampled from (cls.min_scale, cls.max_scale)
+            **kwargs (dict): Any additional keyword-specific arguments required by subclass implementation
+        """
+        # Call super first
+        super().generate_particles(
+            positions=positions,
+            velocities=velocities,
+            orientations=orientations,
+            scales=scales,
+            **kwargs,
+        )
+
+        # Grab pre-existing vels -- note that this already includes the newly included particles, so we will only
+        # keep the first (N - n_new) values
+        current_lin_vels, current_ang_vels = cls.get_particles_velocities()
+
+        # Update the tensors
+        n_particles = len(positions)
+        velocities = np.zeros((n_particles, 3)) if velocities is None else velocities
+        angular_velocities = np.zeros_like(velocities) if angular_velocities is None else angular_velocities
+
+        velocities = np.concatenate([current_lin_vels[:-n_particles], velocities], axis=0)
+        angular_velocities = np.concatenate([current_ang_vels[:-n_particles], angular_velocities], axis=0)
+
+        # Set the vels
+        cls.set_particles_velocities(lin_vels=velocities, ang_vels=angular_velocities)
+
+    @classmethod
+    def create(cls, name, create_particle_template, particle_density, scale, **kwargs):
+        """
+        Utility function to programmatically generate monolithic visual particle system classes.
+
+        Note: If using super() calls in any functions, we have to use slightly esoteric syntax in order to
+        accommodate this procedural method for using super calls
+        cf. https://stackoverflow.com/questions/22403897/what-does-it-mean-by-the-super-object-returned-is-unbound-in-python
+            Use: super(cls).__get__(cls).<METHOD_NAME>(<KWARGS>)
+
+        Note: Scale is automatically inferred from the particle object that is loaded!
```

Comment:
Is this comment accurate? I thought scale is [1, 1, 1] if not specified.

Reply:
Make sure always infer particle radius from (scaled) USD, ideally scale is [1, 1, 1] assuming the asset has the correct scale.

Reply:
done